gcsfs/core.py

```diff
@@ -181,6 +181,94 @@ def validate_response(r, path):
             raise RuntimeError(m)
 
 
+def _dict_to_credentials(token, scopes):
+    """
+    Convert old dict-style token.
+
+    Does not preserve access token itself, assumes refresh required.
+    """
+    try:
+        token = service_account.Credentials.from_service_account_info(
+            token, scopes=scopes
+        )
+    except:  # noqa: E722
+        # TODO: catch specific exceptions
+        token = Credentials(
+            None,
+            refresh_token=token["refresh_token"],
+            client_secret=token["client_secret"],
+            client_id=token["client_id"],
+            token_uri="https://www.googleapis.com/oauth2/v4/token",
+            scopes=scopes,
+        )
+    return token
+
+
+def _connect_token(token, scopes):
+    if isinstance(token, str):
+        if not os.path.exists(token):
+            raise FileNotFoundError(token)
+        try:
+            # is this a "service" token?
+            session, project = _connect_service(token)
+            return session, project
+        except:  # noqa: E722
+            # TODO: catch specific exceptions
+            # some other kind of token file
+            # will raise exception if is not json
+            token = json.load(open(token))
+    if isinstance(token, dict):
+        credentials = _dict_to_credentials(token, scopes)
+    elif isinstance(token, google.auth.credentials.Credentials):
+        credentials = token
+    else:
+        raise ValueError("Token format not understood")
+    session = AuthorizedSession(credentials)
+    return session, None
+
+
+def _connect_google_default(scopes):
+    credentials, project = gauth.default(scopes=scopes)
+    session = AuthorizedSession(credentials)
+    return session, project
+
+
+def _connect_cloud():
+    credentials = gauth.compute_engine.Credentials()
+    session = AuthorizedSession(credentials)
+    return session, None
+
+
+def _connect_service(scopes, fn):
+    # raises exception if file does not match expectation
+    credentials = service_account.Credentials.from_service_account_file(
+        fn, scopes=scopes
+    )
+    session = AuthorizedSession(credentials)
+    return session, None
+
+
+def _connect_cache(project, access, tokens):
+    if (project, access) in tokens:
+        credentials = tokens[(project, access)]
+        session = AuthorizedSession(credentials)
+        return session, None
+
+
+def _connect_anon():
+    return requests.Session(), None
+
+
+def _connect_browser(project, access, scopes):
+    flow = InstalledAppFlow.from_client_config(client_config, scopes)
+    credentials = flow.run_console()
+    # XXX: tokens
+    # self.tokens[(project, access)] = credentials
```

Comment:
This is actually quite important: so someone can authenticate via the browser the first time, but afterwards use the same refresh token for future auths.

Reply:
Oh, I think I got that working. One minute.